File: app/src/test/java/com/futurice/freesound/feature/audio/ExoPlayerAudioPlayerTest.java
/*
 * Copyright 2017 Futurice GmbH
 *                            
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *                            
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.futurice.freesound.feature.audio;

import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.source.MediaSource;

import org.junit.Before;
import org.junit.Test;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.concurrent.TimeUnit;

import io.reactivex.subjects.BehaviorSubject;
import polanski.option.Option;

import static com.futurice.freesound.test.assertion.rx.RxJava2OptionAssertions.isNone;
import static com.futurice.freesound.test.assertion.rx.RxJava2OptionAssertions.isSome;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class ExoPlayerAudioPlayerTest {

    @Mock
    private ExoPlayer exoPlayer;

    @Mock
    private MediaSourceFactory mediaSourceFactory;

    @Mock
    private ObservableExoPlayer observableExoPlayer;

    private ExoPlayerAudioPlayer exoPlayerAudioPlayer;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        exoPlayerAudioPlayer = new ExoPlayerAudioPlayer(exoPlayer,
                                                        observableExoPlayer,
                                                        mediaSourceFactory);
    }

    @Test
    public void stop_stopsExoPlayer() {
        exoPlayerAudioPlayer.stopPlayback();

        verify(exoPlayer).stop();
    }

    @Test
    public void stop_clearsCurrentUrl() {
        PlaybackSource playbackSource = new PlaybackSource(new Id("id"), "url");
        ArrangeBuilder arrangeBuilder = new ArrangeBuilder();
        arrangeBuilder.act()
                      .togglePlayback(playbackSource)
                      .init();
        arrangeBuilder.withPlayingExoPlayer();

        exoPlayerAudioPlayer.stopPlayback();
        arrangeBuilder.withIdleExoPlayer();

        exoPlayerAudioPlayer.getPlayerStateOnceAndStream()
                            .test()
                            .assertValue(isNone(PlayerState::getSource));
    }

    @Test
    public void getPlayerStateStream_initiallyHasNoPlaybackUrl() {
        new ArrangeBuilder()
                .withIdleExoPlayer();

        exoPlayerAudioPlayer.init();

        exoPlayerAudioPlayer.getPlayerStateOnceAndStream()
                            .test()
                            .assertValue(isNone(PlayerState::getSource));
    }

    @Test
    public void toggle_toPlay_playsSource() {
        String id = "id";
        String url = "url";
        PlaybackSource playbackSource = new PlaybackSource(new Id(id), url);
        new ArrangeBuilder()
                .withIdleExoPlayer()
                .withMediaSource()
                .act()
                .init();

        exoPlayerAudioPlayer.togglePlayback(playbackSource);

        verify(mediaSourceFactory).create(url);
        verify(exoPlayer).prepare(any(MediaSource.class));
        verify(exoPlayer).setPlayWhenReady(true);
    }

    @Test
    public void toggle_toPlay_setsPlaybackUrl() {
        String id = "id";
        String url = "url";
        PlaybackSource playbackSource = new PlaybackSource(new Id(id), url);
        ArrangeBuilder arrangeBuilder = new ArrangeBuilder();
        arrangeBuilder.withIdleExoPlayer()
                      .withMediaSource()
                      .act()
                      .init();

        exoPlayerAudioPlayer.togglePlayback(playbackSource);
        arrangeBuilder.withPlayingExoPlayer();

        exoPlayerAudioPlayer.getPlayerStateOnceAndStream()
                            .test()
                            .assertValue(v -> v.getSource().equals(Option.ofObj(playbackSource)));
    }

    @Test
    public void toggle_toPause_pausesSource() {
        String id = "id";
        String url = "url";
        PlaybackSource playbackSource = new PlaybackSource(new Id(id), url);
        ArrangeBuilder arrangeBuilder = new ArrangeBuilder();
        arrangeBuilder.withIdleExoPlayer()
                      .withMediaSource()
                      .act()
                      .init()
                      .togglePlayback(playbackSource);
        arrangeBuilder.withPlayingExoPlayer();

        exoPlayerAudioPlayer.togglePlayback(playbackSource);

        verify(exoPlayer).setPlayWhenReady(false);
    }

    @Test
    public void toggle_toPause_retainsPlaybackUrl() {
        String id = "id";
        String url = "url";
        PlaybackSource playbackSource = new PlaybackSource(new Id(id), url);
        ArrangeBuilder arrangeBuilder = new ArrangeBuilder();
        arrangeBuilder.withIdleExoPlayer()
                      .withMediaSource()
                      .act()
                      .init()
                      .togglePlayback(playbackSource);
        arrangeBuilder.withPlayingExoPlayer();

        exoPlayerAudioPlayer.togglePlayback(playbackSource);
        arrangeBuilder.withPausedExoPlayer();

        exoPlayerAudioPlayer.getPlayerStateOnceAndStream()
                            .test()
                            .assertValue(isSome(PlayerState::getSource));
    }

    @Test
    public void toggle_toUnpause_unpausesSource() {
        String id = "id";
        String url = "url";
        PlaybackSource playbackSource = new PlaybackSource(new Id(id), url);
        ArrangeBuilder arrangeBuilder = new ArrangeBuilder();
        arrangeBuilder.withIdleExoPlayer()
                      .withMediaSource()
                      .act()
                      .init()
                      .togglePlayback(playbackSource);
        arrangeBuilder.withPlayingExoPlayer();
        arrangeBuilder.act()
                      .togglePlayback(playbackSource);
        arrangeBuilder.withPausedExoPlayer();

        exoPlayerAudioPlayer.togglePlayback(playbackSource);

        InOrder inOrder = inOrder(exoPlayer, mediaSourceFactory);
        inOrder.verify(mediaSourceFactory).create(url);
        inOrder.verify(exoPlayer).setPlayWhenReady(true);
        inOrder.verify(exoPlayer).setPlayWhenReady(false);
        inOrder.verify(exoPlayer).setPlayWhenReady(true);
    }

    @Test
    public void toggle_doesNotClearPlaybackUrl_whenPausing() {
        String id = "id";
        String url = "url";
        PlaybackSource playbackSource = new PlaybackSource(new Id(id), url);
        ArrangeBuilder arrangeBuilder = new ArrangeBuilder()
                .withIdleExoPlayer()
                .withMediaSource();
        arrangeBuilder.act()
                      .init()
                      .togglePlayback(playbackSource);
        arrangeBuilder.withPlayingExoPlayer();

        exoPlayerAudioPlayer.togglePlayback(playbackSource);
        arrangeBuilder.withPausedExoPlayer();

        exoPlayerAudioPlayer.getPlayerStateOnceAndStream()
                            .test()
                            .assertValue(isSome(PlayerState::getSource));
    }

    @Test
    public void toggle_withNewUrl_playsNewSource_whenPlaying() {
        String id1 = "id1";
        String id2 = "id2";
        String url1 = "url1";
        String url2 = "url2";
        PlaybackSource playbackSource1 = new PlaybackSource(new Id(id1), url1);
        PlaybackSource playbackSource2 = new PlaybackSource(new Id(id2), url2);
        ArrangeBuilder arrangeBuilder = new ArrangeBuilder()
                .withIdleExoPlayer()
                .withMediaSource();
        arrangeBuilder.act()
                      .init()
                      .togglePlayback(playbackSource1);
        arrangeBuilder.withPlayingExoPlayer();

        exoPlayerAudioPlayer.togglePlayback(playbackSource2);

        InOrder inOrder = inOrder(exoPlayer, mediaSourceFactory);
        inOrder.verify(mediaSourceFactory).create(url1);
        inOrder.verify(exoPlayer).prepare(any(MediaSource.class));
        inOrder.verify(exoPlayer).setPlayWhenReady(true);
        inOrder.verify(mediaSourceFactory).create(url2);
        inOrder.verify(exoPlayer).prepare(any(MediaSource.class));
        inOrder.verify(exoPlayer).setPlayWhenReady(true);
    }

    @Test
    public void toggle_withNewUrl_playsNewSource_whenEnded() {
        String id1 = "id1";
        String id2 = "id2";
        String url1 = "url1";
        String url2 = "url2";
        PlaybackSource playbackSource1 = new PlaybackSource(new Id(id1), url1);
        PlaybackSource playbackSource2 = new PlaybackSource(new Id(id2), url2);
        ArrangeBuilder arrangeBuilder = new ArrangeBuilder()
                .withIdleExoPlayer()
                .withMediaSource();
        arrangeBuilder.act()
                      .init()
                      .togglePlayback(playbackSource1);
        arrangeBuilder.withEndedExoPlayer();

        exoPlayerAudioPlayer.togglePlayback(playbackSource2);

        InOrder inOrder = inOrder(exoPlayer, mediaSourceFactory);
        inOrder.verify(mediaSourceFactory).create(url1);
        inOrder.verify(exoPlayer).prepare(any(MediaSource.class));
        inOrder.verify(exoPlayer).setPlayWhenReady(true);
        inOrder.verify(mediaSourceFactory).create(url2);
        inOrder.verify(exoPlayer).prepare(any(MediaSource.class));
        inOrder.verify(exoPlayer).setPlayWhenReady(true);
    }

    @Test
    public void release_releasesExoPlayer() {
        exoPlayerAudioPlayer.release();

        verify(exoPlayer).release();
    }

    private class ArrangeBuilder {

        private final BehaviorSubject<ExoPlayerState> exoPlayerStateStream = BehaviorSubject
                .create();
        private final BehaviorSubject<Long> exoPlayerProgressStream = BehaviorSubject.create();

        ArrangeBuilder() {
            when(observableExoPlayer.getExoPlayerStateOnceAndStream())
                    .thenReturn(exoPlayerStateStream);
            when(observableExoPlayer.getTimePositionMsOnceAndStream(1, TimeUnit.SECONDS))
                    .thenReturn(exoPlayerProgressStream);
        }

        ArrangeBuilder withMediaSource() {
            when(mediaSourceFactory.create(anyString())).thenReturn(mock(MediaSource.class));
            return this;
        }

        ArrangeBuilder withIdleExoPlayer() {
            return withExoPlayerStateStreamEvent(false, ExoPlayer.STATE_IDLE);
        }

        ArrangeBuilder withPlayingExoPlayer() {
            return withExoPlayerStateStreamEvent(true, ExoPlayer.STATE_READY);
        }

        ArrangeBuilder withPausedExoPlayer() {
            return withExoPlayerStateStreamEvent(false, ExoPlayer.STATE_READY);
        }

        ArrangeBuilder withEndedExoPlayer() {
            return withExoPlayerStateStreamEvent(true, ExoPlayer.STATE_ENDED);
        }

        ArrangeBuilder withExoPlayerStateStreamEvent(boolean playWhenReady, int state) {
            exoPlayerStateStream.onNext(new ExoPlayerState(playWhenReady, state));
            return this;
        }

        Act act() {
            return new Act();
        }

    }

    private class Act {

        Act init() {
            exoPlayerAudioPlayer.init();
            return this;
        }

        Act togglePlayback(PlaybackSource playbackSource) {
            exoPlayerAudioPlayer.togglePlayback(playbackSource);
            return this;
        }

    }

}

File: app/src/test/java/com/futurice/freesound/feature/audio/ExoPlayerAudioPlayerTest.kt
/*
 * Copyright 2017 Futurice GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.futurice.freesound.feature.audio

import com.futurice.freesound.test.rx.TimeSkipScheduler
import com.futurice.freesound.test.rx.TrampolineSchedulerProvider
import io.reactivex.Scheduler
import io.reactivex.observers.TestObserver
import io.reactivex.schedulers.TestScheduler
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import java.util.concurrent.TimeUnit

class ExoPlayerAudioPlayerTest {

    private lateinit var schedulerProvider: TrampolineSchedulerProvider
    private lateinit var fakeExoPlayer: FakeObservableExoPlayer
    private lateinit var exoPlayerAudioPlayer: ExoPlayerAudioPlayer

    private val updatePeriod = 100L
    private val updateUnits = TimeUnit.MILLISECONDS

    private fun TestScheduler.tick() {
        advanceTimeBy(updatePeriod, updateUnits)
        triggerActions()
    }

    @Before
    fun setUp() {
        schedulerProvider = TrampolineSchedulerProvider()
        fakeExoPlayer = FakeObservableExoPlayer()
        exoPlayerAudioPlayer = ExoPlayerAudioPlayer(fakeExoPlayer,
                updatePeriod,
                updateUnits,
                schedulerProvider)
    }

    @Test
    fun `playerState is Idle when ExoPlayer is Idle`() {
        // given
        exoPlayerAudioPlayer.init()

        // when, then
        exoPlayerAudioPlayer.playerStateOnceAndStream
                .test()
                .assertValue { state -> state is PlayerState.Idle }
    }

    @Test
    fun `togglePlayback follows playback sequence from Idle`() {
        // given
        val timerScheduler = TestScheduler()
        val playbackSource1 = PlaybackSource(Id("abc"), "url1")
        val playbackSource2 = PlaybackSource(Id("def"), "url2")
        ArrangeBuilder()
                .withTimeScheduler(timerScheduler)
        exoPlayerAudioPlayer.init()
        val states: TestObserver<in PlayerState> = exoPlayerAudioPlayer.playerStateOnceAndStream.test()

        // when
        exoPlayerAudioPlayer.togglePlayback(playbackSource1) // play
        fakeExoPlayer.setProgress(500)
        timerScheduler.tick()
        exoPlayerAudioPlayer.togglePlayback(playbackSource1) // pause
        exoPlayerAudioPlayer.togglePlayback(playbackSource1) // resume
        fakeExoPlayer.setProgress(1000)
        timerScheduler.tick()
        exoPlayerAudioPlayer.togglePlayback(playbackSource2) // play different

        // then
        states.assertValues(PlayerState.Idle,
                PlayerState.Assigned(playbackSource1, PlaybackStatus.PLAYING, 0),
                PlayerState.Assigned(playbackSource1, PlaybackStatus.PLAYING, 500),
                PlayerState.Assigned(playbackSource1, PlaybackStatus.PAUSED, 500),
                PlayerState.Assigned(playbackSource1, PlaybackStatus.PLAYING, 500),
                PlayerState.Assigned(playbackSource1, PlaybackStatus.PLAYING, 1000),
                PlayerState.Assigned(playbackSource2, PlaybackStatus.PLAYING, 0))
    }

    @Test
    fun `togglePlayback plays URL when is Idle`() {
        // given
        val playbackSource = PlaybackSource(Id("abc"), "url")
        ArrangeBuilder()
                .withTimeScheduler(TestScheduler())
        exoPlayerAudioPlayer.init()

        // when
        exoPlayerAudioPlayer.togglePlayback(playbackSource)

        // then
        exoPlayerAudioPlayer.playerStateOnceAndStream.test()
                .assertValue {
                    it == PlayerState.Assigned(playbackSource,
                            PlaybackStatus.PLAYING, 0)
                }
    }

    @Test
    fun `togglePlayback plays URL when is Ended`() {
        // given
        val playbackSource = PlaybackSource(Id("abc"), "url")
        ArrangeBuilder()
                .withTimeScheduler(TestScheduler())
                .withEndedExoPlayer(playbackSource)

        // when
        exoPlayerAudioPlayer.togglePlayback(playbackSource)

        // then
        exoPlayerAudioPlayer.playerStateOnceAndStream.test()
                .assertValue {
                    it == PlayerState.Assigned(playbackSource,
                            PlaybackStatus.PLAYING, 0)
                }
    }

    @Test
    fun `togglePlayback pauses when same URL is Playing`() {
        // given
        val playbackSource = PlaybackSource(Id("abc"), "url")
        ArrangeBuilder()
                .withTimeScheduler(TestScheduler())
                .withPlayingExoPlayer(playbackSource)

        // when
        exoPlayerAudioPlayer.togglePlayback(playbackSource)

        // then
        exoPlayerAudioPlayer.playerStateOnceAndStream.test()
                .assertValue {
                    it == PlayerState.Assigned(playbackSource,
                            PlaybackStatus.PAUSED, 0)
                }
    }

    @Test
    fun `togglePlayback resumes URL when same URL is Paused`() {
        // given
        val playbackSource = PlaybackSource(Id("abc"), "url")
        ArrangeBuilder()
                .withTimeScheduler(TestScheduler())
                .withPausedExoPlayer(playbackSource)

        // when
        exoPlayerAudioPlayer.togglePlayback(playbackSource)

        // then
        exoPlayerAudioPlayer.playerStateOnceAndStream.test()
                .assertValue {
                    it == PlayerState.Assigned(playbackSource,
                            PlaybackStatus.PLAYING, 0)
                }
    }

    @Test
    fun `togglePlayback plays new URL when different URL is Playing`() {
        // given
        val playbackSource1 = PlaybackSource(Id("abc"), "url1")
        val playbackSource2 = PlaybackSource(Id("def"), "url2")
        ArrangeBuilder()
                .withTimeScheduler(TestScheduler())
                .withPlayingExoPlayer(playbackSource1)

        // when
        exoPlayerAudioPlayer.togglePlayback(playbackSource2)

        // then
        exoPlayerAudioPlayer.playerStateOnceAndStream.test()
                .assertValue {
                    it == PlayerState.Assigned(playbackSource2,
                            PlaybackStatus.PLAYING, 0)
                }
    }

    @Test
    fun `togglePlayback plays URL different URL is Paused`() {
        // given
        val playbackSource1 = PlaybackSource(Id("abc"), "url1")
        val playbackSource2 = PlaybackSource(Id("def"), "url2")
        ArrangeBuilder()
                .withTimeScheduler(TestScheduler())
                .withPausedExoPlayer(playbackSource1)

        // when
        exoPlayerAudioPlayer.togglePlayback(playbackSource2)

        // then
        exoPlayerAudioPlayer.playerStateOnceAndStream.test()
                .assertValue {
                    it == PlayerState.Assigned(playbackSource2,
                            PlaybackStatus.PLAYING, 0)
                }
    }

    @Test
    fun `togglePlayback plays URL different URL is Ended`() {
        // given
        val playbackSource1 = PlaybackSource(Id("abc"), "url1")
        val playbackSource2 = PlaybackSource(Id("def"), "url2")
        ArrangeBuilder()
                .withTimeScheduler(TestScheduler())
                .withEndedExoPlayer(playbackSource1)

        // when
        exoPlayerAudioPlayer.togglePlayback(playbackSource2)

        // then
        exoPlayerAudioPlayer.playerStateOnceAndStream.test()
                .assertValue {
                    it == PlayerState.Assigned(playbackSource2,
                            PlaybackStatus.PLAYING, 0)
                }
    }

    @Test
    fun `stopPlayback stops ExoPlayer`() {
        // given
        ArrangeBuilder()
                .withPlayingExoPlayer(PlaybackSource(Id("abc"), "url"))

        // when
        exoPlayerAudioPlayer.stopPlayback()

        // then
        exoPlayerAudioPlayer.playerStateOnceAndStream
                .test()
                .assertValue { it is PlayerState.Idle }
    }

    @Test
    fun `release releases ExoPlayer`() {
        // given, when
        exoPlayerAudioPlayer.release()

        // then
        assertThat(fakeExoPlayer.isReleased).isTrue()
    }

    private inner class ArrangeBuilder() {

        init {
            withTimeSkipScheduler()
        }

        fun withPlayingExoPlayer(playbackSource: PlaybackSource): ArrangeBuilder {
            exoPlayerAudioPlayer.init()
            exoPlayerAudioPlayer.togglePlayback(playbackSource)
            return this;
        }

        fun withPausedExoPlayer(playbackSource: PlaybackSource): ArrangeBuilder {
            exoPlayerAudioPlayer.init()
            exoPlayerAudioPlayer.togglePlayback(playbackSource) // make play
            exoPlayerAudioPlayer.togglePlayback(playbackSource) // make pause
            return this;
        }

        fun withEndedExoPlayer(playbackSource: PlaybackSource): ArrangeBuilder {
            exoPlayerAudioPlayer.init()
            exoPlayerAudioPlayer.togglePlayback(playbackSource) // make play
            fakeExoPlayer.end() // make end
            return this;
        }

        fun withTimeScheduler(scheduler: Scheduler): ArrangeBuilder {
            schedulerProvider.setTimeScheduler({ scheduler })
            return this
        }

        fun withTimeSkipScheduler(): ArrangeBuilder {
            return withTimeScheduler(TimeSkipScheduler.instance())
        }

    }

}
